Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,571 @@
+
+/-- If `E / F` is purely inseparable, then for any reduced ring `L`, the map `(E →+* L) → (F →+* L)`
+induced by `algebraMap F E` is injective. In other words, a purely inseparable field extension
+is an epimorphism in the category of fields. -/
+theorem IsPurelyInseparable.injective_comp_algebraMap [h : IsPurelyInseparable F E]
+    (L : Type w) [CommRing L] [IsReduced L] :
+    Function.Injective fun f : E →+* L ↦ f.comp (algebraMap F E) := fun f g heq ↦ by
+  ext x
+  obtain ⟨q, hF⟩ := ExpChar.exists F
+  obtain ⟨n, y, h⟩ := (isPurelyInseparable_iff_mem_pow F E q).1 h x
+  replace heq := congr($heq y)
+  simp_rw [RingHom.comp_apply, h] at heq
+  cases' hF with _ _ hprime _
+  · rwa [one_pow, pow_one] at heq
+  nontriviality L
+  haveI := charP_of_injective_ringHom (f.comp (algebraMap F E)).injective q
+  haveI := Fact.mk hprime
+  simp_rw [map_pow, ← iterate_frobenius] at heq
+  exact iterate_frobenius_inj L q n heq
+
+-- TODO:
+/- If `L` is an algebraically closed field containing `E`, such that the map
+`(E →+* L) → (F →+* L)` induced by `algebraMap F E` is injective, then `E / F` is
+purely inseparable. In other words, epimorphisms in the category of fields must be
+purely inseparable extensions. -/
+-- theorem IsPurelyInseparable.of_injective_comp_algebraMap (L : Type w) [Field L] [IsAlgClosed L]
+--     (hn : Nonempty (E →+* L)) (h : Function.Injective fun f : E →+* L ↦ f.comp (algebraMap F E)) :
+--     IsPurelyInseparable F E := by
+--   sorry
+
+end IsPurelyInseparable
+
+namespace IntermediateField
+
+instance isPurelyInseparable_bot : IsPurelyInseparable F (⊥ : IntermediateField F E) :=
+  (botEquiv F E).symm.isPurelyInseparable
+
+/-- `F⟮x⟯ / F` is a purely inseparable extension if and only if the mininal polynomial of `x`
+has separable degree one. -/
+theorem isPurelyInseparable_adjoin_simple_iff_natSepDegree_eq_one {x : E} :
+    IsPurelyInseparable F F⟮x⟯ ↔ (minpoly F x).natSepDegree = 1 := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · have := (isPurelyInseparable_iff_natSepDegree_eq_one F _).1 h (AdjoinSimple.gen F x)
+    rwa [minpoly_eq] at this
+  have hx : IsIntegral F x := by_contra fun h' ↦ by
+    simp only [minpoly.eq_zero h', natSepDegree_zero, zero_ne_one] at h
+  haveI := adjoin.finiteDimensional hx
+  rwa [isPurelyInseparable_iff_finSepDegree_eq_one _ _ (Algebra.IsAlgebraic.of_finite F F⟮x⟯),
+    finSepDegree_adjoin_simple_eq_natSepDegree _ _ hx.isAlgebraic]
+
+/-- If `F` is of exponential characteristic `q`, then `F⟮x⟯ / F` is a purely inseparable extension
+if and only if `x ^ (q ^ n)` is contained in `F` for some `n : ℕ`. -/
+theorem isPurelyInseparable_adjoin_simple_iff_mem_pow (q : ℕ) [hF : ExpChar F q] {x : E} :
+    IsPurelyInseparable F F⟮x⟯ ↔ ∃ n : ℕ, x ^ q ^ n ∈ (algebraMap F E).range := by
+  simp_rw [isPurelyInseparable_adjoin_simple_iff_natSepDegree_eq_one,
+    minpoly.natSepDegree_eq_one_iff_mem_pow q]
+
+private theorem isPurelyInseparable_adjoin_finset_of_mem_pow (q : ℕ) [hF : ExpChar F q]
```

Comment:
Why is this private? I see no harm in it being public.

Reply:
Same for the other private lemmas below.

Reply:
Because it is a special case of `isPurelyInseparable_adjoin_iff_mem_pow.2`. Same as the following private lemmas.

Reply:
You can still keep it public and/or add a pointer to `isPurelyInseparable_adjoin_iff_mem_pow` in the docstring.